### ISSUE TITLE
Unhandled Exception during decoding

### ISSUE THREAD

**Issue by ChristopherHaws** [sabnzbd/sabnzbd#642]
Hello,

My SABnzbd instance keeps crashing on me several times per day. When looking at the logs, it appears to be crashing due to an unhandled exception in yEnc during decoding.

Last log entry before the application crashes:
```
2016-07-30 20:09:51,915::INFO::[assembler:88] Decoding \\?\M:\Downloads\Incomplete\file.part099.rar yenc
```

This is what is getting logged in the Event Viewer:
```
Faulting application name: SABnzbd.exe, version: 1.0.3.0, time stamp: 0x49180193
Faulting module name: _yenc.pyd, version: 0.0.0.0, time stamp: 0x564a4896
Exception code: 0xc0000005
Fault offset: 0x00001648
Faulting process id: 0x1b54
Faulting application start time: 0x01d1ead6d4b106cc
Faulting application path: C:\Program Files (x86)\SABnzbd\SABnzbd.exe
Faulting module path: C:\Program Files (x86)\SABnzbd\lib\_yenc.pyd
Report Id: 45d3b4ec-56cc-11e6-80c2-00155d01ba05
Faulting package full name: 
Faulting package-relative application ID: 
```

I am using:
SABnzbd: 1.0.3 [927ba3c]
Windows: Server 2012 R2 [6.3.9600]

**Comment by shypike** [comment]
No idea really. Yenc is a third-party library.
We never get reports like this from Windows users.
One thing we do recommend against is having a external drive as temporary download folder.
Not sure whether that relates to the yEnc crash.

**Comment by thezoggy** [comment]
may want to run memtest / disable any antivirus or malware type thing.. something that does DEP

**Comment by Safihre** [comment]
Some system problem seems more likely indeed, the yEnc has been there for years and years with 0 issues.

**Comment by ChristopherHaws** [comment]
I don't have antivirus running on the machine and memory is all fine (I ran a memtest a few weeks ago). What you said about external drives might be on the right track. I am using Windows Storage Spaces with a volume using the new ReFS file system as my temporary download location. Tonight I will try making a new volume that is NTFS and see if its an issue with the new ReFS file system, and if that doesn't fix the problem I will try moving the temp directory to the local C drive. I will let you know if either of these things fix the problem.

Thanks!

**Comment by thezoggy** [comment]
@Safihre yes exactly, it'd be safe to assume the issue is with the users machine/setup than yEnc due to how stable its been over the years. why i suggested memory/antivirus/etc stuff that could mess with it...

**Comment by GregSawin** [comment]
The same thing has started happening for me as well.  I've been running years without issue and then starting a few weeks ago it crashes every few days with the same event log error as ChristopherHaws.  The only recent change I can think of is the Win10 anniversary update.  I'm using version 1.0.3 [927ba3c].

**Comment by Safihre** [comment]
Could you check for us if the same happens on 0.7.20? 
Just to test and see if it's Windows update or Sab, the yEnc module didn't change between versions. 
https://github.com/sabnzbd/sabnzbd/releases/tag/0.7.20

**Comment by kevatkinson** [comment]
Adding a +1 to this, Sab has worked for 18 months + on a dedicated PC I have for it, but I've noticed it start exiting whilst downloading. 

Faulting application name: SABnzbd.exe, version: 1.0.3.0, time stamp: 0x49180193
Faulting module name: _yenc.pyd, version: 0.0.0.0, time stamp: 0x564a4896
Exception code: 0xc0000005
Fault offset: 0x00001648
Faulting process ID: 0x9c8
Faulting application start time: 0x01d20d1a9edf81dd
Faulting application path: C:\Program Files (x86)\SABnzbd\SABnzbd.exe
Faulting module path: C:\Program Files (x86)\SABnzbd\lib\_yenc.pyd
Report ID: 52139033-3ae8-4b12-ab3f-605b4781c047
Faulting package full name: 
Faulting package-relative application ID: 

Windows 10 x64 on a 120GB SSD with 8GB RAM (2x 4GB). All NTFS, using Windows Defender. Also on 1.0.3 [927ba3c]

**Comment by albino1** [comment]
1.0.3 is a couple of versions old now.  Maybe try upgrading.

**Comment by ChristopherHaws** [comment]
@albino1 I am on 1.1.0 and this issue is still happening for me.

**Comment by thezoggy** [comment]
can you post logs? also how are you running sab.. from source? binary build?

**Comment by shypike** [comment]
Looking at the log, it's a binary build and the third-party yenc library is the culprit.
Does this happen with specific NZB files or randomly?
When it crashes like this and you re-download the complete NZB, does it crash again?

**Comment by thezoggy** [comment]
@shypike you use yenc 0.3 or 0.4 in the binary builds?

**Comment by Safihre** [comment]
Could any of you maybe send us an NZB on which this happens? Safihre@sabnzbd.org

**Comment by ChristopherHaws** [comment]
Hello,

I downloaded the 0.4 yenc binaries from @thezoggy and I have not had a crash since (its been about 1 week). Today I tried switching the binaries back to the 0.3 version and had a crash within 20 minutes. I will switch back to the 0.4 binaries and let you know if I have any issues, but I haven't had any yet.

Thanks!

**Comment by thezoggy** [comment]
@ChristopherHaws thats great to hear. Perhaps shy will include the updated yenc with future builds...

**Comment by Safihre** [comment]
Interesting! What's the difference between 0.3 and 0.4 actually? Is there any difference in the function that we use?

**Comment by thezoggy** [comment]
@Safihre 

```
version: 0.4
- change license to LGPL
- crc32 codes are now Python longs (returned from _yenc calls) in the range [2**0, 2**32 -1]
- crc32 formatting is fixed
- 'bytes' arguments named 'bytez' to avoid collision with 'bytes' type
- basic unit testing
- unescaped dots at the beginning of line caused ydecode to fail under some circumstances
  dots at the beginning of line are now always escaped 
- Encoder's and Decoder's output files (if any) are automatically flushed and closed upon deletion
- new method close() on Encoder and Decoder
```

then my version has the win 'hacks' applied so you can compile it on windows.. then built out for 32 and 64bit python

**Comment by Safihre** [comment]
@thezoggy On a side note, I have been working on an improved version that will do all the steps we now do in Python from raw data to decoded yEnc in new C module, based on_yenc. I finished it last week and it's 42% faster :) Integrating it now into Sab, more work than I thought to keep backwards compatibility!

**Comment by shypike** [comment]
@thezoggy of course I will.

**Comment by Safihre** [comment]
@ChristopherHaws @kevatkinson 
We just released 1.1.1RC1 which includes yEnc 0.4, can you let us know if this fixes it?
https://sabnzbd.org/downloads

**Comment by ChristopherHaws** [comment]
Sure. I will install it now and let it run for a day or two and let you know how it goes.

Thanks!

**Comment by thezoggy** [comment]
just to note, this may be why yenc 0.3.x was crashing for certain people:
https://fedorapeople.org/~dmalcolm/gcc-python-plugin/2012-03-05/python-yenc-0.3-10.fc17/

which was addressed in 0.4.x
`_yenc.c: Don't crash if malloc/PyString_*() fails (out of memory)`
